debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/OracleErrorHandler.java

```diff
@@ -7,42 +7,80 @@
+
+        // Contents of this list should be any type of error message text
+        // The error check uses case-insensitive contains semantics
+        retryOracleMessageContainsTexts.add("No more data to read from socket");
+    }
+
     public OracleErrorHandler(String logicalName, ChangeEventQueue<?> queue) {
         super(OracleConnector.class, logicalName, queue);
     }
 
     @Override
     protected boolean isRetriable(Throwable throwable) {
-        if (throwable.getMessage() == null || throwable.getCause() == null) {
-            return false;
-        }
+        while (throwable != null) {
+            // Always retry any recoverable error
+            if (throwable instanceof SQLRecoverableException) {
+                return true;
+            }
+
+            // If message is provided, run checks against it
+            final String message = throwable.getMessage();
+            if (message != null && message.length() > 0) {
+                // Check Oracle error codes
+                for (String errorCode : retryOracleErrors) {
+                    if (message.startsWith(errorCode)) {
+                        return true;
+                    }
+                }
+                // Check Oracle error message texts
+                for (String messageText : retryOracleMessageContainsTexts) {
+                    if (message.toUpperCase().contains(messageText.toUpperCase())) {
+                        return true;
+                    }
+                }
+            }
 
-        return throwable.getMessage().startsWith("ORA-03135") || // connection lost
-                throwable.getMessage().startsWith("ORA-12543") || // TNS:destination host unreachable
-                throwable.getMessage().startsWith("ORA-00604") || // error occurred at recursive SQL level 1
-                throwable.getMessage().startsWith("ORA-01089") || // Oracle immediate shutdown in progress
-                throwable.getMessage().startsWith("ORA-01333") || // Failed to establish LogMiner dictionary
-                throwable.getMessage().startsWith("ORA-01284") || // Redo/Archive log cannot be opened, likely locked
-                throwable.getMessage().startsWith("ORA-26653") || // Apply DBZXOUT did not start properly and is currently in state INITIALI
-                throwable.getMessage().startsWith("ORA-01291") || // missing logfile
-                throwable.getMessage().startsWith("ORA-01327") || // failed to exclusively lock system dictionary as required BUILD
-                throwable.getMessage().startsWith("ORA-04030") || // out of process memory
-                throwable.getCause() instanceof IOException ||
-                throwable instanceof SQLRecoverableException ||
-                throwable.getMessage().toUpperCase().contains("NO MORE DATA TO READ FROM SOCKET") ||
-                (throwable.getCause() != null && throwable.getCause().getCause() instanceof NetException);
+            if (throwable.getCause() != null) {
```

Comment:
I must admit that I am not sure the code aligns with the comment. It seems to me from the comment that this condition should be checked only once, for top level, not for any level in the stack. If it is the case than it should be placed out of the `while` loop.

Reply:
I don't think it is particularly wise to assume a position in the stack trace.  As our code evolves and new try/catch blocks may get added, that could easily shift.  It's a bit difficult for us to implement such a test so that we know we've broken something.  Furthermore, it's possible that the position may change with the driver's evolution as well.

In addition, the same could be said for the `IOException` check which only checked if the top level's cause.  

So in being prepared and flexible, I felt it was best that we actually simply check the entire stack trace for such a condition instead.  If you'd prefer I revert back I can but I think the more flexible route is the best approach.

Reply:
I am perfectly fine with that, just then it make probably sense. Re-reading comment in the light of the explanantion it is probabaly valid as well.